File: src/modules/udp.py
import struct
from socket import socket
from socket import AF_INET
from socket import SOCK_DGRAM

class SwitchRes():
    pixel_clock: float
    hactive: int
    hbegin: int
    hend: int
    htotal: int
    vactive: int
    vbegin: int
    vend: int
    vtotal: int
    refresh_rate: int

    def __init__(self, modeline: str = "", refresh_rate: int = 60) -> None:
        self.pixel_clock = 4.905
        self.hactive = 256
        self.hbegin = 264
        self.hend = 287
        self.htotal = 312
        self.vactive = 240
        self.vbegin = 241
        self.vend = 244
        self.vtotal = 262
        self.refresh_rate= refresh_rate
        self._parse_modeline(modeline)
    
    def _parse_modeline(self, modeline: str) -> None:
        parts = modeline.split(" ")
        if modeline == "" or len(parts) != 9:
            return
        self.pixel_clock = float(parts[0])
        self.hactive = int(parts[1])
        self.hbegin = int(parts[2])
        self.hend = int(parts[3])
        self.htotal = int(parts[4])
        self.vactive = int(parts[5])
        self.vbegin = int(parts[6])
        self.vend = int(parts[7])
        self.vtotal = int(parts[8])
        
        


class UdpClient():
    UDP_IP: str
    UDP_PORT: int
    MTU_BLOCK_SIZE = 1470

    sock: socket
    width: int
    height: int
    frame: int

    def __init__(self, udp_port: int = 32100, udp_ip: str = '127.0.0.1'):
        self.sock = socket(AF_INET, # Internet
            SOCK_DGRAM) # UDP        
        #self.sock.bind((UDP_IP, UDP_PORT))
        self.UDP_IP = udp_ip
        self.UDP_PORT = udp_port
        self.width = 256
        self.height = 240
        self.frame = 0

    def send_packet(self, packet: bytearray):
        self.sock.sendto(packet, (self.UDP_IP, self.UDP_PORT))

    def cmd_init(self):
        buffer: bytearray = bytearray(4)
        struct.pack_into('<B', buffer, 0, 2) # CMD
        struct.pack_into('<B', buffer, 1, 0) # lz4 compression flag
        struct.pack_into('<B', buffer, 2, 0) # sound rate flag
        struct.pack_into('<B', buffer, 3, 0) # sound channel
        self.send_packet(buffer)
        #print([i for i in buffer])

    def cmd_close(self):
        buffer: bytearray = bytearray(1)
        struct.pack_into('>B', buffer, 0, 1) # CMD
        self.send_packet(buffer)
        #print([i for i in buffer])

    def cmd_switchres(self, switchres: SwitchRes):
        self.width = switchres.hactive
        self.height = switchres.vactive
        buffer: bytearray = bytearray(26)
        struct.pack_into('<B', buffer, 0, 3) # CMD
        struct.pack_into('<d', buffer, 1, switchres.pixel_clock)
        struct.pack_into('<H', buffer, 9, switchres.hactive)
        struct.pack_into('<H', buffer, 11, switchres.hbegin)
        struct.pack_into('<H', buffer, 13, switchres.hend)
        struct.pack_into('<H', buffer, 15, switchres.htotal)
        struct.pack_into('<H', buffer, 17, switchres.vactive)
        struct.pack_into('<H', buffer, 19, switchres.vbegin)
        struct.pack_into('<H', buffer, 21, switchres.vend)
        struct.pack_into('<H', buffer, 23, switchres.vtotal)
        struct.pack_into('<B', buffer, 25, 0) # interlace
        self.send_packet(buffer)
        #print([i for i in buffer])

    def cmd_blit(self, frame_buffer: bytearray):
        self.frame = self.frame + 1
        buffer: bytearray = bytearray(9)
        struct.pack_into('<B', buffer, 0, 6) # CMD
        struct.pack_into('<I', buffer, 1, self.frame)
        struct.pack_into('<H', buffer, 5, 0) # vsyncAuto
        struct.pack_into('<B', buffer, 7, 0) # lz4 blockSize & 0xff
        struct.pack_into('<B', buffer, 8, 0) # lz4 blockSize >> 8
        self.send_packet(buffer)
        #print([i for i in buffer])
        self.send_mtu(frame_buffer)
    
    def send_mtu(self, buffer: bytearray):
        bytes_to_send = len(buffer)
        chunk_max_size = self.MTU_BLOCK_SIZE
        chunk_size: int = 0
        offset: int = 0
        while bytes_to_send > 0:
            chunk_size = chunk_max_size if bytes_to_send > chunk_max_size else bytes_to_send
            bytes_to_send = bytes_to_send - chunk_size
            self.send_packet(buffer[offset : offset+chunk_size])
            #print([i for i in buffer[offset : offset+chunk_size]])
            offset += chunk_size

    def wait_for_ack(self):
        # Untested Placeholder
        while True:
            data, addr = self.sock.recvfrom(self.MTU_BLOCK_SIZE)
            print("received ack")

File: pkg/groovymister/udp.go
package groovymister

import (
	"encoding/binary"
	"fmt"
	"math"
	"net"
)

const (
	cmdHeaderClose     byte = 1
	cmdHeaderInit      byte = 2
	cmdHeaderSwitchRes byte = 3
	cmdHeaderBlit      byte = 6
)

type UdpClient struct {
	host         string
	conn         net.PacketConn
	addr         *net.UDPAddr
	iConn        net.PacketConn
	iAddr        *net.UDPAddr
	frame        uint32
	mtuBlockSize int32
}

func (client *UdpClient) SendPacket(buffer []byte) {
	//fmt.Println("Sending Packet length", len(buffer))
	_, err := client.conn.WriteTo(buffer, client.addr)
	if err != nil {
		panic(err)
	}
}

func (client *UdpClient) SendMTU(buffer []byte) {
	//fmt.Println(buffer[3000:3100])
	bytesToSend := int32(len(buffer))
	//fmt.Println(bytesToSend, client.mtuBlockSize)
	chunkMaxSize := int32(client.mtuBlockSize)
	var chunkSize int32 = 0
	var offset int32 = 0
	for bytesToSend > 0 {
		chunkSize = chunkMaxSize
		if bytesToSend <= chunkMaxSize {
			chunkSize = bytesToSend
		}
		bytesToSend = bytesToSend - chunkSize
		client.SendPacket(buffer[offset : offset+chunkSize])
		offset += chunkSize
	}
}

func (client *UdpClient) CmdClose() {
	buffer := make([]byte, 1)
	buffer[0] = cmdHeaderClose
	client.SendPacket(buffer)
	client.Close()
}

func (client *UdpClient) CmdInit() {
	buffer := make([]byte, 5)
	buffer[0] = cmdHeaderInit
	buffer[1] = 0 // lz4 compression flag
	buffer[2] = 0 // sound rate
	buffer[3] = 0 // sound channel
	buffer[4] = 0 // rgb mode
	client.SendPacket(buffer)
}

func (client *UdpClient) CmdSwitchres(modeline *Modeline) {
	buffer := make([]byte, 26)
	buffer[0] = cmdHeaderSwitchRes
	binary.LittleEndian.PutUint64(buffer[1:9], math.Float64bits(modeline.PixelClock))
	binary.LittleEndian.PutUint16(buffer[9:11], modeline.HActive)
	binary.LittleEndian.PutUint16(buffer[11:13], modeline.HBegin)
	binary.LittleEndian.PutUint16(buffer[13:15], modeline.HEnd)
	binary.LittleEndian.PutUint16(buffer[15:17], modeline.HTotal)
	binary.LittleEndian.PutUint16(buffer[17:19], modeline.VActive)
	binary.LittleEndian.PutUint16(buffer[19:21], modeline.VBegin)
	binary.LittleEndian.PutUint16(buffer[21:23], modeline.VEnd)
	binary.LittleEndian.PutUint16(buffer[23:25], modeline.VTotal)
	buffer[25] = 0
	if modeline.Interlace {
		buffer[25] = 1
	}
	client.SendPacket(buffer)
}

func (client *UdpClient) CmdBlit(frameBuffer []byte) {
	client.frame++
	buffer := make([]byte, 7)
	buffer[0] = cmdHeaderBlit
	binary.LittleEndian.PutUint32(buffer[1:5], client.frame)
	binary.LittleEndian.PutUint16(buffer[5:7], 0) // vsyncAuto
	//buffer[7] = 0                                 // lz4 blockSize & 0xff
	//buffer[8] = 0                                 // lz4 blockSize >> 8
	client.SendPacket(buffer)
	//start := time.Now()
	client.SendMTU(frameBuffer)
	//fmt.Println("blit took", time.Since(start))
}

func (client *UdpClient) Close() {
	client.conn.Close()
	client.iConn.Close()
}

func (client *UdpClient) PollInput() (chan GroovyInputPacket, chan bool) {
	inputChan := make(chan GroovyInputPacket, 20)
	inputQuitChan := make(chan bool, 1)
	go func() {
		prevPacket := GroovyInputPacket{}
		for {
			select {
			case <-inputQuitChan:
				return
			default:
				//fmt.Println("listening for input")
				buf := make([]byte, 1024)
				rlen, _, err := client.iConn.ReadFrom(buf)
				//fmt.Println("UDP READ", rlen)
				if err != nil {
					fmt.Println("UDP READ ERROR", err)
				}
				if rlen == 9 {
					newPacket := InputPacketFromBuffer(buf[:rlen])
					if newPacket.JoyFrame > prevPacket.JoyFrame || (newPacket.JoyFrame == prevPacket.JoyFrame && newPacket.JoyOrder > prevPacket.JoyOrder) {
						prevPacket = newPacket
						inputChan <- newPacket
					}
				}
			}
		}
	}()
	return inputChan, inputQuitChan
}

func NewUdpClient(host string, mtuBlockSize int32) UdpClient {
	var client UdpClient
	client.host = host
	conn, err := net.ListenPacket("udp4", ":32110")
	if err != nil {
		panic(err)
	}
	addr, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("%s:32100", host))
	if err != nil {
		panic(err)
	}
	iConn, err := net.ListenPacket("udp4", ":32111")
	if err != nil {
		panic(err)
	}
	iAddr, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("%s:32101", host))
	if err != nil {
		panic(err)
	}
	// send empty?
	_, _ = iConn.WriteTo(make([]byte, 1), iAddr)

	client.conn = conn
	client.addr = addr
	client.iConn = iConn
	client.iAddr = iAddr
	client.mtuBlockSize = mtuBlockSize
	return client
}
